### ISSUE TITLE
Any Interest in an L5 GPS Version?

### ISSUE THREAD

**Issue by AgileEngineeringLLC** [taroz/GNSS-SDRLIB#26]
I have gnss-sdrlib running just fine for L1-only gps. Is anyone interested in collaborating to create an L5-only version? The L5 acq and track algorithms from the CU-SDR (Matlab) project could be leveraged as a start. 

My setup for gnss-sdrlib is an i86 processor running Ubuntu LTS, BladeRF 2.0 front end. I also have a 3-channel Firehose front end, but there is no real-time API built for it yet. Running on an M1/M2 Mac could be interesting too, but would need to get the needed Arm-based libs running.

**Comment by ZiglioUK** [comment]
Hi Don, I'm keen. I've known gnss-sdrlib for many years and occasionally go back to eat. It's nice and simple and small but even less robust than gnss-sdr.
Why are you so keen on gnss-sdrlib? I think there's space for a new/smaller gnss engine than gnss-sdr, maybe written in Rust

**Comment by AgileEngineeringLLC** [comment]
I think gnss-sdrlib may be the quickest way to go. Seems like we'd just need to add L5 channels, using the L1 channels and the CU-SDR L5 algorithms as the template. But Rust seems very possible as well.

But here are a few thoughts:

Rust. Looks really promising, but would require a complete re-write of the gnss-sdrlib functions. And I'm unsure if the libraries exist or are robust (FFT, TCP, real-time buffering, etc).

GNSS-SDR. Very robust, but to me seems very bloated. Also, the real-time engine is GnuRadio's buffering scheme, and it's very complex and tough to figure out. A lot of OO bloat to me in GNSS-SDR.

GNSS-SDRLIB. Sleek, simple design, very real-time functional based. Relatively easy to follow the threading approach. Rust might be good at this as well!

CU-SDR Collection. Has a great implementation of Matlab L5-only tracking, but is not channelized for real-time (if converted to C or Rust). Firehose L5 data processes great, but BladeRF L5 data does not for some reason (at least for me).

Firehose vs. Blade RF Front Ends. Firehose L5 IQ files playback great in CU-SDR, bladeRF L5 files do not. No real-time API for Firehose yet, but could be done with TCP buffering.

FieldTrip C Buffer. Looks like a great C library for TCP buffering, may work well with getting Firehose to real-time.

Don

**Comment by AgileEngineeringLLC** [comment]
An article on some challenges of using Rust: 

(https://mdwdotla.medium.com/using-rust-at-a-startup-a-cautionary-tale-42ab823d9454)

**Comment by ZiglioUK** [comment]
Thanks @AgileEngineeringLLC , that article isabout using rust for a web app / cloud service, it's not quite the same space.
Anyway, after working a lot with gnss-sdr and gnss-sdrlib here are my observations:

1. gnss-sdr is a text book implementation of a gnss receiver that lacks robustness. Some of the problems are: fragile acquisition and tracking, complex configuration, 3rd party navigation engine (rtk-lib, poorly integrated)

2. gnss-sdr lib is a much a smaller project that hasn't been maintained for many years. Its beauty is in its simplicity. In my opinion it's even less robust than gnss-sdr because for example, from memory, it doesn't even check for loss of trackiing.

From the performance point of view gnss-sdr should be superior because it builds on top of optimized DSP kernels, ie volk.
In terms of fixing either, it's not simple, I'm not sure if gnss-sdrlib would be a good base to start with but adding L5 shouldn't be too hard. I've added Galileo myself to another engine, and that's quite similar.

Do you maintain a branch of gnss-sdrlib that compiles?

Sorry, I've never heard of CU-SDR or Firehose, I'll check them out.
I've got a HackRF but no L5 antennas at the moment.

**Comment by AgileEngineeringLLC** [comment]
I have a couple of HackRFs, really like them, but they only sample reliably to about 10 Msps in my experience, and are older USB so very slow interface unfortunately. The Firehose is a terrific ethernet 3-channel RF front end, but no real-time API yet, and are about $800 US. The bladeRFs are nice, but i can't seem to get them to track on L5.

What i like about gnss-sdrlib is its simplicity, and its very readable code. So seems like adding L5 and increasing robustness are very doable. But i really like the idea of Rust, although that feels like a pretty long-term conversion project to me. 

I maintain my own versions of compiled gnss-sdrlib myself in dropbox. I download via github, but have't tried maintaining my own versions there yet... need to learn that. 

Perhaps an interesting challenge would be to convert CU-SDR from matlab to Rust. Then once that works in file playback mode, channelize it for real-time (a challenge for sure) and add a real-time API / buffer. Longer term project too, but would have fully working CU-SDR functions to debug against.

**Comment by AgileEngineeringLLC** [comment]
https://transitiva.com/product/gnss-firehose-receiver-tri-band-quad-constellation/

**Comment by ZiglioUK** [comment]
Hi Don, apologies for the delay: I've just started a new job and I'm finishing a course on Coursera. I've forked this repo and will start working on L5 in my spare time.
There are example captures in the gnss-sdr repo if I'm not mistaken

**Comment by AgileEngineeringLLC** [comment]
Congrats on the new job, that's terrific!

I can load a Firehose set of IQ binary files (L1, L2, L5) in Dropbox if that'd be helpful. I was able to read Firehose files just fine in CU-SDR, and all three modes (L1-only, L2-only, and L5-only) process all the way through to PVT solution just fine.

I tried to get gnss-sdrlib to read in a Firehose L1 file as a test case, and was having some issues. It's sampled at 69.984 Msps and is in int8 format, so seems like it should be sort of easy. I'm getting acquisition, but not nav decoding / tracking, and getting memory allocation issue. 

In the mean time, I've started porting CU-SDR over to Octave, and then I thought I'd use OctaveCoder to port it to C++. If that all works, I'd re-look at what it'd take to channelize it for real-time (maybe too big of an effort) like gnss-sdr does.
